base/src/main/java/uk/ac/ebi/atlas/search/analyticsindex/FullAnalyticsSearchService.java

```diff
@@ -0,0 +1,251 @@
+        this.restTemplate = restTemplate;
+        this.avgExpressionOverBioentityIdentifiersFacet = avgExpressionOverBioentityIdentifiersFacet;
+        this.maxExpressionOverBioentityIdentifiersFacet = maxExpressionOverBioentityIdentifiersFacet;
+    }
+
+    public List<String> searchSpecificExpression(String experimentAccession,
+                                                 Double expressionThreshold,
+                                                 Set<String> assayGroupIds) {
+        return searchSpecificExpression(experimentAccession, expressionThreshold, assayGroupIds, DEFAULT_RESULT_SIZE);
+    }
+
+    public List<String> searchSpecificExpression(String experimentAccession,
+                                                 Double expressionThreshold,
+                                                 Set<String> assayGroupIds,
+                                                 int resultSize) {
+
+        String filterQueries = MessageFormat.format(FQ_TEMPLATE, experimentAccession);
+
+        SolrQuery solrQuery = new SolrQuery();
+        solrQuery.setRows(0)
+                .setFilterQueries(filterQueries)
+                .setQuery(
+                        MessageFormat.format(SELECTED_ASSAY_GROUP_IDS_Q_TEMPLATE, Joiner.on(",").join(assayGroupIds), expressionThreshold))
+                .set("json.facet", ResourceUtils.readPlainTextResource(avgExpressionOverBioentityIdentifiersFacet));
+
+        Stopwatch stopwatch1 = Stopwatch.createStarted();
+        LOGGER.debug(
+                "Searching for genes in {} {} across {} assay groups...",
+                experimentAccession, expressionThreshold, assayGroupIds.size());
+
+        TreeMultimap<Integer, Pair<String, Double>> genesBySpecificity =
+                parseAverageExpressionOverSelectedAssayGroups(
+                        fetchResponseAsString("http://localhost:8983/solr/analytics/query", solrQuery), resultSize);
+
+        LOGGER.debug("Finished in {} ms", stopwatch1.elapsed(TimeUnit.MILLISECONDS));
+
+        Map<String, Double> avgExpressionByGeneId = genesBySpecificity.keySet().stream()
+                .flatMap(specificity -> genesBySpecificity.get(specificity).stream())
+                .collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
+
+
+        String nonSelectedQuery = avgExpressionByGeneId.keySet().size() > TERMS_QUERY_SIZE_THRESHOLD ?
+                MessageFormat.format(
+                        NON_SELECTED_ASSAY_GROUP_IDS_Q_TEMPLATE_ALT,
+                        Joiner.on(",").join(assayGroupIds),
+                        expressionThreshold) :
+                MessageFormat.format(
+                        NON_SELECTED_ASSAY_GROUP_IDS_Q_TEMPLATE,
+                        Joiner.on(",").join(assayGroupIds),
+                        Joiner.on(",").join(avgExpressionByGeneId.keySet()),
+                        expressionThreshold);
+
+        solrQuery.setQuery(nonSelectedQuery)
```

Comment:
I found reusing solrQuery a bit surprising.

Reply:
It’s ugly, right? 😆